sympy/solvers/ode.py

```diff
@@ -2432,6 +2432,30 @@ def checkodesol(ode, sol, func=None, order='auto', solve_for_func=True):
 
             if sol.lhs == func:
                 s = sub_func_doit(ode_diff, func, sol.rhs)
+                if s.has(besselj):
+                    while True:
+                        besseljs = s.find(lambda x: isinstance(x, besselj))
+                        for ba in sorted(besseljs, key=lambda x: re(x.args[0])):
+                            a, x = ba.args
+                            bap1 = besselj(a+1, x)
+                            bap2 = besselj(a+2, x)
+                            if s.has(bap1) and s.has(bap2):
+                                s = s.subs(ba, 2*(a+1)/x*bap1 - bap2)
+                                break
+                        else:
+                            break
+                if s.has(bessely):
+                    while True:
+                        besselys = s.find(lambda x: isinstance(x, bessely))
+                        for ba in sorted(besselys, key=lambda x: re(x.args[0])):
+                            a, x = ba.args
+                            bap1 = bessely(a+1, x)
+                            bap2 = bessely(a+2, x)
+                            if s.has(bap1) and s.has(bap2):
+                                s = s.subs(ba, 2*(a+1)/x*bap1 - bap2)
+                                break
+                        else:
+                            break
```

Comment:
Move all the code above into a separate function called e.g. _bessel_simp_recursion. The in that function define another function that you call like `_use_recursion(besselj)` So you don't duplicate the same code for besselj and bessely.

Reply:
okk

Reply:
```
diff --git a/sympy/solvers/ode.py b/sympy/solvers/ode.py
index 07efc6a..2597893 100644
--- a/sympy/solvers/ode.py
+++ b/sympy/solvers/ode.py
@@ -2326,6 +2326,28 @@ def _expand(expr):
     if len(eq) == 1:
         eq = eq[0]
     return eq
+def _bessel_simp_recursion(expr):
+
+    def _use_recursion(bessel, expr):
+        while True:
+            bessels = expr.find(lambda x: isinstance(x, bessel))
+            try:
+                for ba in sorted(bessels, key=lambda x: re(x.args[0])):
+                    a, x = ba.args
+                    bap1 = bessel(a+1, x)
+                    bap2 = bessel(a+2, x)
+                    if expr.has(bap1) and expr.has(bap2):
+                        expr = expr.subs(ba, 2*(a+1)/x*bap1 - bap2)
+                        break
+                else:
+                    return expr
+            except (ValueError, TypeError):
+                return expr
+    if expr.has(besselj):
+        expr = _use_recursion(besselj, expr)
+    if expr.has(bessely):
+        expr = _use_recursion(bessely, expr)
+    return expr

```

Reply:
Yes that's what I was thinking. I wonder if there's a reasonable way to include this in besselsimp and use that in checkodesol:
https://github.com/sympy/sympy/blob/6f2860aa633d2822003a289166a4d5b17e42ca51/sympy/simplify/simplify.py#L1104

Reply:
I tried  to implement in `besselsimp ` function and it works fine
-
```
diff --git a/sympy/simplify/simplify.py b/sympy/simplify/simplify.py
index 1593e32..f67f4cb 100644
--- a/sympy/simplify/simplify.py
+++ b/sympy/simplify/simplify.py
@@ -11,7 +11,7 @@
 from sympy.core.numbers import Float, I, pi, Rational, Integer
 from sympy.core.rules import Transform
 from sympy.core.sympify import _sympify
-from sympy.functions import gamma, exp, sqrt, log, exp_polar, piecewise_fold
+from sympy.functions import gamma, exp, sqrt, log, exp_polar, piecewise_fold, re
 from sympy.functions.combinatorial.factorials import CombinatorialFunction
 from sympy.functions.elementary.complexes import unpolarify
 from sympy.functions.elementary.exponential import ExpBase
@@ -1183,6 +1183,30 @@ def repl(nu, z):
     expr = expr.replace(besseli, expander(besseli))
     expr = expr.replace(besselk, expander(besselk))
 
+    def _bessel_simp_recursion(expr):
+
+        def _use_recursion(bessel, expr):
+            while True:
+                bessels = expr.find(lambda x: isinstance(x, bessel))
+                try:
+                    for ba in sorted(bessels, key=lambda x: re(x.args[0])):
+                        a, x = ba.args
+                        bap1 = bessel(a+1, x)
+                        bap2 = bessel(a+2, x)
+                        if expr.has(bap1) and expr.has(bap2):
+                            expr = expr.subs(ba, 2*(a+1)/x*bap1 - bap2)
+                            break
+                    else:
+                        return expr
+                except (ValueError, TypeError):
+                    return expr
+        if expr.has(besselj):
+            expr = _use_recursion(besselj, expr)
+        if expr.has(bessely):
+            expr = _use_recursion(bessely, expr)
+        return expr
+
+    expr = _bessel_simp_recursion(expr)
     if expr != orig_expr:
         expr = expr.factor()

```

Reply:
https://github.com/sympy/sympy/blob/6f2860aa633d2822003a289166a4d5b17e42ca51/sympy/simplify/simplify.py#L1130

for this I have made a function with the help of `bessel_simp`

```
[44]:         def _use_relation(expr): 
    ...:             while True: 
    ...:                 bessels = expr.find(lambda x: isinstance(x, besselj)) 
    ...:                 try: 
    ...:                     for ba in bessels: 
    ...:                         a, x = ba.args 
    ...:                         bap1 = besselj(-a, x) 
    ...:                         if expr.has(ba*cos(pi*a)) and expr.has(bap1): 
    ...:                             expr = expr.subs(ba*cos(pi*a), sin(pi*a)*bessely(a, x)+bap1) 
    ...:                             break 
    ...:                     else: 
    ...:                         return expr 
    ...:                 except (ValueError, TypeError): 
    ...:                     return expr
```

Reply:
what do you think about this?

Reply:
I think we'd need to look carefully at what besselsimp already does and see why it doesn't already work for this and whether this is a reasonable addition to its existing function.

Reply:
> I think we'd need to look carefully at what besselsimp already does and see why it doesn't already work for this and whether this is a reasonable addition to its existing function.

Any suggestions?

Reply:
Have you looked at besselsimp to see what it does currently? It looks like it already does something similar to this for other kinds of bessel functions...

We shouldn't modify that function unless we are sure that what we are doing is consistent with what it already does. Otherwise we should just leave it as it is and add this code to checkodesol.

Reply:
I think this can be part of besselsimp.

Can you check exactly which functions satisfy this recurrence relation?

Also we need to work out whether this should be before or after the other simplifications that are already there.

Reply:
The existing besselsimp routine already uses this recurrence for bessel functions of integer order:
```
In [13]: besselsimp(besselj(2, x))                                              
Out[13]: 
-(x⋅besselj(0, x) - 2⋅besselj(1, x)) 
─────────────────────────────────────
                  x 
```
The reason that doesn't work for the cases we are interested in is that the order is non-integer e.g. 1+I. Currently besselsimp always acts to reduce the order for integer order leading to an expression in terms of bessel functions of order 1 and 0 as above.

The code you've added here substitutes for higher order i.e. it rewrites besselj(a, x) in terms of besselj(a+1, x) and besselj(a+2, x). Maybe it should work the other way and try to rewrite in terms of lower orders. I'm not sure how these simplifications will interact in different situations.

Reply:
> https://github.com/sympy/sympy/blob/6f2860aa633d2822003a289166a4d5b17e42ca51/sympy/simplify/simplify.py#L1130
> 
> for this, I have made a function with the help of `bessel_simp`
> 
> ```
> [44]:         def _use_relation(expr): 
>     ...:             while True: 
>     ...:                 bessels = expr.find(lambda x: isinstance(x, besselj)) 
>     ...:                  
>     ...:                     for ba in bessels: 
>     ...:                         a, x = ba.args 
>     ...:                         bap1 = besselj(-a, x) 
>     ...:                         if expr.has(ba*cos(pi*a)) and expr.has(bap1): 
>     ...:                             expr = expr.subs(ba*cos(pi*a), sin(pi*a)*bessely(a, x)+bap1) 
>     ...:                             break 
>     ...:                     else: 
>     ...:                         return expr 
>     ...:                 
>     ...:                    
> ```
> 
> I can add this also to` besselsimp` function

I think we can add this recurrence relation also because it is in "to do work" and not implemented yet

Reply:
Can you explain exactly what the recurrence is? We need to ensure that all of these recursions are converging on a uniquely defined form for the expression.

Reply:
In to_do we have to implement -
![image](https://user-images.githubusercontent.com/37741324/60073038-f671dd00-973c-11e9-8289-d013e05a8582.png)

So first I tried to find `besselj(n,x)` and `besselj(-n,x)` and then check if an expression has `besselj(n,x).cos()` and `besselj(-n,x)` if true then I replace `besselj(n,x).cos()` to `sin()*bessely(n, x)+besselj(-n,x)`.
Using this recurrence relation we remove division and find a link between `besselj` and `bessely`

Reply:
@oscarbenjamin any suggestion?

Reply:
It seems unlikely to me that this particular recurrence will arise in practice so that this simplification will be generally useful but I might be wrong. I think that for this PR it's best to stick to the recurrences that we know will be useful.

I'm going away for a while now so won't be able to follow up here. Hopefully others can guide you with this.

Reply:
I don't think that the formula defining Neumann's function Y_alpha is a recursion formula. If alpha is not an integer, then J_alpha and J_{-alpha} are linearly independent solutions of the Bessel equation. As the equation is of order two, all solutions, Y_alpha included, can be represented as a linear combination of them.

If alpha is an integer n, the solutions J_n and J_{-n} are either the same or opposite, depending on the parity on n. A linearly independent second solution can be obtained in this case by taking the difference `J_alpha*cos(pi*alpha) - J_{-alpha}` and dividing it by `sin(pi*alpha)` to get a non-zero result as alpha tends to n.

In summary: If alpha is not an integer, then the solutions of the Bessel equation can be represented by means of J_alpha and J_{-alpha} and Y_alpha is not needed. If alpha is an integer, the formula defining Y_alpha becomes 0/0 and cannot be used.

Reply:
> In summary: If alpha is not an integer, then the solutions of the Bessel equation can be represented by means of J_alpha and J_{-alpha} and Y_alpha is not needed. If alpha is an integer, the formula defining Y_alpha becomes 0/0 and cannot be used.


I got you and I will not add this relation.
Thanks